### PR TITLE
Harden vs. TaskState collisions

### PR DESCRIPTION
**Pull request by crusaderky**
#6525 removed the assertion that multiple TaskState objects can't exist at the same time, because it was incompatible with the new instances weakset.
This PR reintroduces the check, but limited to the WorkerState and more methodical. It also makes sure that validate_state *will* fail if the previous incarnation of a task remained in the WorkerState for any reason, whereas before it would only happen in case of key hash collision.

### CODE REVIEW

**Comment by crusaderky** [comment]
This PR accidentally relates with #6585, which is dealing with a very similar design issue on the scheduler side:
- the `SchedulerState` holding sets of `scheduler.WorkerState` objects there
- the `worker_state_machine.WorkerState` holding sets of `TaskState` objects here

In both cases, there *should* never be a case of duplicate objects in the set but we can't enforce it through a key.
We should agree on a coherent design for the two.

**Review by fjetter — CHANGES_REQUESTED** [review]
We should not merge this until the conversation in https://github.com/dask/distributed/pull/6585#issuecomment-1161504498 is settled.

File: distributed/worker_state_machine.py [hunk 1]
```diff
@@ -283,7 +283,8 @@ def __eq__(self, other: object) -> bool:
         return other is self
 
     def __hash__(self) -> int:
-        return hash(self.key)
+        # See note in __eq__
+        return id(self)
```
[the remark this answers is not in the document]
Comment (fjetter) [reply distributed/worker_state_machine.py:287]: Is this change required to get the validation passing?

[the remark this answers is not in the document]
Comment (gjoseph92) [reply distributed/worker_state_machine.py:287]: Strictly speaking, both `__eq__` and `__hash__` are unnecessary to get the behavior we want:

> User-defined classes have [`__eq__()`](https://docs.python.org/3/reference/datamodel.html#object.__eq__) and [`__hash__()`](https://docs.python.org/3/reference/datamodel.html#object.__hash__) methods by default; with them, all objects compare unequal (except with themselves) and `x.__hash__()` returns an appropriate value such that `x == y` implies both that `x is y` and `hash(x) == hash(y)`.
https://docs.python.org/3/reference/datamodel.html#object.__hash__

I'd kinda prefer to not implement them at all and just use the defaults. We can keep a comment explaining this, but having the explicit methods feels redundant.

**Pull request merged**